Add support for EAP Dump in RADIUS Messages

aravindprasad90 opened the pull request:
Usually, for EAP related authentications, RADIUS messages contain EAP payload. Currently, RADIUS dump messages donot display the EAP frame information. Hence, added changes for displaying EAP info within RADIUS dump messages. 

Sample O/P:

1) EAP Request Identity Frame:

Before this change:
```
         EAP-Message Attribute (79), length: 19, Value: .
```
       
After this change:
```
           EAP-Message Attribute (79), length: 10, Value: Response (2), id 1, len 8
                 Type Identity (1), Identity: avi
            0x0000:  0201 0008 0161 7669
```

2) EAP Request Challenge Frame:
Before this change:
```
          EAP-Message Attribute (79), length: 24, Value: ..
            0x0000:  0102 0016 0410 cd32 576a 04bd ea97 14c6
            0x0010:  9e84 ae7c 6de2
```

After this change:
```
          EAP-Message Attribute (79), length: 24, Value: Request (1), id 2, len 22
                 Type MD5-challenge (4)
            0x0000:  0102 0016 0410 cd32 576a 04bd ea97 14c6
            0x0010:  9e84 ae7c 6de2
```

3) EAP Response Challenge frame:

Before this change:
```
          EAP-Message Attribute (79), length: 27, Value: ..
            0x0000:  0202 0019 0410 035f 1f16 c4e2 d71e 1553
            0x0010:  6b55 effc 0a0f 6176 69
```

After this change:
```
          EAP-Message Attribute (79), length: 27, Value: Response (2), id 2, len 25
                 Type MD5-challenge (4)
            0x0000:  0202 0019 0410 035f 1f16 c4e2 d71e 1553
            0x0010:  6b55 effc 0a0f 6176 69
```

4) EAP Success Frame:

Before this change:
```
          EAP-Message Attribute (79), length: 6, Value: ..
            0x0000:  0302 0004
```

After this change:
```
          EAP-Message Attribute (79), length: 6, Value: Success (3), id 2, len 4
            0x0000:  0302 0004
```

fxlb commented:
For getting data from a packet buffer, use the new GET_ macros instead of EXTRACT_ ones (See 176e182 and ee68aa3).

aravindprasad90 commented:
@fxlb Thanks for the review. Replaced all the EXTRACT macros with the GET_ macros. 

Kindly review and let me know your views.

aravindprasad90 commented:
@fxlb Thanks for the review. I have addressed the review comment. Kindly review and let me know your views.

fxlb commented on print-eap.c:
```diff
@@ -148,56 +148,36 @@ static const struct tok eap_aka_subtype_values[] = {
-
-            switch (subtype) {
+    u_int type, subtype;
+    int len, count;
```
Should `len` be unsigned?
(line 183: `len = GET_BE_U_2(tptr + 2);`)

aravindprasad90 replied:
The eap_print() API had the "len" attribute as signed previously and hence, didnot want to disturb it and retained the same.  

But, as you pointed out, it should have been unsigned in the first place. Hence, changed it to unsigned now. Thanks.

fxlb commented on print-eap.c:
```diff
@@ -148,56 +148,36 @@ static const struct tok eap_aka_subtype_values[] = {
+    type = GET_U_1(tptr);
+    ND_TCHECK_2(tptr + 2);
+    len = GET_BE_U_2(tptr + 2);
+    if(len != (int)length) {
```
Get rid of `(int)` cast?

aravindprasad90 replied:
Thanks, addressed

fxlb commented on print-eap.c:
```diff
@@ -148,56 +148,36 @@ static const struct tok eap_aka_subtype_values[] = {
+    ND_TCHECK_2(tptr + 2);
+    len = GET_BE_U_2(tptr + 2);
+    if(len != (int)length) {
+       goto trunc; 
```
Trailing space to remove.

aravindprasad90 replied:
Thanks, addressed

fxlb commented on tests/radius-v.out:
```diff
@@ -9,7 +9,8 @@
 	  Service-Type Attribute (6), length: 6, Value: Framed
 	  Framed-MTU Attribute (12), length: 6, Value: 1500
-	  EAP-Message Attribute (79), length: 19, Value: .
+	  EAP-Message Attribute (79), length: 19, Value: , Response (2), id 0, len 17
```
Why `Value: ` (empty)

aravindprasad90 replied:
Actually, the EAP Attribute is in itself a TLV and hence, the value cannot be printed as such like a string/number. 
Hence, previously . was printed as a result. 

Now, with the new changes, the TLV is extracted and contents are printed. I previously wanted to separate out the extracted part and the value part and hence, value was empty.

But, as you pointed out, it makes more sense to avoid printing empty value and print the extracted contents directly.
Addressed the same.

Thanks a lot for the extensive review.

fxlb commented on tests/radius-v.out:
```diff
@@ -18,7 +19,8 @@
 	  Service-Type Attribute (6), length: 6, Value: Framed
 	  Reply-Message Attribute (18), length: 11, Value: Hello, %u
-	  EAP-Message Attribute (79), length: 24, Value: ..
+	  EAP-Message Attribute (79), length: 24, Value: , Request (1), id 1, len 22
```
Same. There are others.

aravindprasad90 replied:
Addressed. Thanks

fxlb commented on print-eap.c:
```diff
@@ -148,56 +148,36 @@ static const struct tok eap_aka_subtype_values[] = {
 void
 eap_print(netdissect_options *ndo,
-          const u_char *cp,
+          const u_char *tptr,
```
Why changing `cp` to `tptr`?
in eapol_print(), you use `cp`.
`tptr` should be reserved as temporary copy (if needed?) in a function body, not parameter.

aravindprasad90 replied:
Agree with your comments, Addressed. Thanks

aravindprasad90 commented:
@fxlb Big thanks for your time and extensive review. I have addressed all the review comments. 

Kindly review and let me know your views. Thanks.

fxlb commented:
I have added a verbose output for the test case eapon1.pcap.
When the PR in rebased on top of master, it shows a unwanted change.
```
32c32
<    14  15:17:35.622870 EAP packet (0) v1, len 5, Request (1), id 1, len 5
---
>    14  15:17:35.622870 EAP packet (0) v1, len 5Request (1), id 1, len 5
```
See : https://travis-ci.org/github/fxlb/tcpdump/jobs/709917622
Please, rebase the eap-radius-support branch on top of master and fix.

fxlb commented on print-eap.c:
```diff
@@ -278,10 +258,46 @@ eap_print(netdissect_options *ndo,
+    u_int eap_type, tlen, eap_len;
+
+    ndo->ndo_protocol = "eap";
+    tptr = cp;
```
Is tptr needed? just use cp?

aravindprasad90 replied:
Agreed, cp is sufficient. Removed tptr.

fxlb commented on print-eap.c:
```diff
@@ -278,10 +258,46 @@ eap_print(netdissect_options *ndo,
+
+    ndo->ndo_protocol = "eap";
+    tptr = cp;
+    tlen = length;
```
Is tlen needed? just use length?

aravindprasad90 replied:
Agreed, removed tlen. Thanks.

aravindprasad90 commented:
> I have added a verbose output for the test case eapon1.pcap.
> When the PR is rebased on top of master, it shows a unwanted change.
> 
> ```
> 32c32
> <    14  15:17:35.622870 EAP packet (0) v1, len 5, Request (1), id 1, len 5
> ---
> >    14  15:17:35.622870 EAP packet (0) v1, len 5Request (1), id 1, len 5
> ```
> 
> See : https://travis-ci.org/github/fxlb/tcpdump/jobs/709917622
> Please, rebase the eap-radius-support branch on top of master and fix.

Hi @fxlb 

Fixed the issue. Kindly let me know if its fine now. 

Thanks for your time and the detailed review again.

fxlb commented:
Rebased, merged, thank you for your patience!

aravindprasad90 commented:
@fxlb A Big thanks for your time and detailed review. Thanks for the support :-).

infrastation commented:
I was going to resume reviewing this proposed feature (this started around 2018, I guess), but got dragged away by other things, please excuse me. Glad you have eventually sorted it.

aravindprasad90 commented:
Hi Denis @infrastation, 

Thanks for the initial reviews and the support.
